python/tritonserver/_api/_datautils.py

```diff
@@ -0,0 +1,753 @@
+            32,
+        ): DataType.FP32,
+        (
+            _dlpack.DLDataTypeCode.kDLFloat,
+            64,
+        ): DataType.FP64,
+        (
+            _dlpack.DLDataTypeCode.kDLBfloat,
+            16,
+        ): DataType.BF16,
+    },
+)
+
+TRITON_TO_DLPACK_DTYPE: dict[DataType, _dlpack.DLDataType] = _CustomKeyErrorDict(
+    "Triton server data type",
+    "DLPack data type",
+    {
+        value: _dlpack.DLDataType(type_code=key[0], bits=key[1], lanes=1)
+        for key, value in DLPACK_TO_TRITON_DTYPE.items()
+    },
+)
+
+NUMPY_TO_TRITON_DTYPE: dict[type, DataType] = _CustomKeyErrorDict(
+    "Numpy data type",
+    "Triton server data type",
+    {
+        bool: DataType.BOOL,
+        numpy.int8: DataType.INT8,
+        numpy.int16: DataType.INT16,
+        numpy.int32: DataType.INT32,
+        numpy.int64: DataType.INT64,
+        numpy.uint8: DataType.UINT8,
+        numpy.uint16: DataType.UINT16,
+        numpy.uint32: DataType.UINT32,
+        numpy.uint64: DataType.UINT64,
+        numpy.float16: DataType.FP16,
+        numpy.float32: DataType.FP32,
+        numpy.float64: DataType.FP64,
+        numpy.bytes_: DataType.BYTES,
+        numpy.str_: DataType.BYTES,
+        numpy.object_: DataType.BYTES,
+    },
+)
+
+TRITON_TO_NUMPY_DTYPE: dict[DataType, type] = _CustomKeyErrorDict(
+    "Triton data type",
+    "Numpy data type",
+    {
+        **{value: key for key, value in NUMPY_TO_TRITON_DTYPE.items()},
+        **{DataType.BYTES: numpy.object_},
+    },
+)
+
+
+@dataclass
+class MemoryBuffer:
+    """Memory allocated for a Tensor"""
+
+    data_ptr: int
+    memory_type: MemoryType
+    memory_type_id: int
+    size: int
+    owner: Any
+
+    @staticmethod
+    def from_dlpack(
+        owner: Any, dlpack_object: Optional[DLPackObject] = None
+    ) -> MemoryBuffer:
+        if dlpack_object is None:
+            if not hasattr(owner, "__dlpack__"):
+                raise InvalidArgumentError("Object does not support DLpack protocol")
+
+            dlpack_object = DLPackObject(owner)
+
+        if not dlpack_object.contiguous:
+            raise InvalidArgumentError("Only contiguous memory is supported")
+
+        return MemoryBuffer(
+            int(dlpack_object.data_ptr),
+            dlpack_object.memory_type,
+            dlpack_object.memory_type_id,
+            dlpack_object.byte_size,
+            owner,
+        )
+
+    def _create_TRITONSERVER_BufferAttributes(self) -> BufferAttributes:
+        buffer_attributes = BufferAttributes()
+        buffer_attributes.memory_type = self.memory_type
+        buffer_attributes.memory_type_id = self.memory_type_id
+        buffer_attributes.byte_size = self.size
+        #        buffer_attributes.cuda_ipc_handle = None
+        return buffer_attributes
+
+
+class MemoryAllocator(ABC):
+    @abstractmethod
+    def allocate(
+        self, size: int, memory_type: MemoryType, memory_type_id: int, tensor_name: str
```

Comment:
I think name is probably not required for the `allocate` function.

Reply:
@rmccorm4 - what do you think - I'm on the fence here - I haven't seen a good case for knowing the tensor name in allocate (maybe for tracking?) - but had kept it since it comes from the underlying response allocator request. We could also remove it and add it back if needed ...

Reply:
I would remove the name for now to keep it simple, they appear unused. Agreed we can add it later if needed.